[pytorch][perf] add speed benchmark binary for torch jit

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25230 [pytorch][perf] add speed benchmark binary for torch jit**
* #25229 [pytorch] fix binaries build for BUILD_CAFFE2_MOBILE=OFF

Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
[pytorch][perf] add speed benchmark binary for torch jit

Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Reviewers: dreiss

Subscribers: pytorch-mobile-dev

Tasks:

Tags:

supriyar reviewed (changes requested):
Added some comments. 
Also I noticed that when I use the `./build_android.sh` script to build this along with the flags, it doesn't build caffe2 speed_benchmark correctly for android. Not sure if we want to be able to build both binaries with same command or not.

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

ljk53 committed:
Update on "[pytorch][perf] add speed benchmark binary for torch jit"


Summary:
Add similar benchmark binary as caffe2 speed benchmark for torch JIT.

Test Plan:
- Test with ResNet-18 torch script model:
```
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --print_output=true
./speed_benchmark_torch --model=res18.pb --input_dims="1,3,224,224" --input_type=float --warmup=5 --iter=20
```

- Verified building as desktop/server binary works:
```
BUILD_BINARY=ON python setup.py develop
```

- Verified building as android binary works:
```
./scripts/build_android.sh \
-DBUILD_BINARY=ON \
-DBUILD_CAFFE2_MOBILE=OFF \
-DCMAKE_PREFIX_PATH=$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())') \
-DPYTHON_EXECUTABLE=$(python -c 'import sys; print(sys.executable)')
```

Differential Revision: [D17067216](https://our.internmc.facebook.com/intern/diff/D17067216)

dzhulgakov commented:
Sorry for coming in late, but it'd have been nice to use this class instead: https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/throughput_benchmark.h - that way we could use the same tool to run multithreaded benchmarks as well.

ezyang commented:
This patchset broke your master android builds: https://circleci.com/gh/pytorch/pytorch/2633580?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

```
Aug 30 03:09:27 ../lib/libc10.so: undefined reference to `__android_log_print'
Aug 30 03:09:27 ../lib/libtorch.so: undefined reference to `__android_log_vprint'
Aug 30 03:09:27 ../lib/libtorch.so: undefined reference to `__android_log_write'
Aug 30 03:09:27 clang++: error: linker command failed with exit code 1 (use -v to see invocation)
Aug 30 03:09:27 binaries/CMakeFiles/run_plan.dir/build.make:97: recipe for target 'bin/run_plan' failed
Aug 30 03:09:27 make[2]: *** [bin/run_plan] Error 1
Aug 30 03:09:27 CMakeFiles/Makefile2:1765: recipe for target 'binaries/CMakeFiles/run_plan.dir/all' failed
Aug 30 03:09:27 make[1]: *** [binaries/CMakeFiles/run_plan.dir/all] Error 2
Aug 30 03:09:28 1 warning generated.
Aug 30 03:09:28 [ 97%] Linking CXX executable ../bin/speed_benchmark_torch
Aug 30 03:09:28 /opt/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: warning: liblog.so, needed by ../lib/libtorch.so, not found (try using -rpath or -rpath-link)
Aug 30 03:09:28 ../lib/libc10.so: undefined reference to `__android_log_print'
Aug 30 03:09:28 ../lib/libtorch.so: undefined reference to `__android_log_vprint'
Aug 30 03:09:28 ../lib/libtorch.so: undefined reference to `__android_log_write'
Aug 30 03:09:28 clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```
